Common/Data/LeanDataWriter.cs

```diff
@@ -172,340 +214,178 @@ public void DownloadAndSave(IBrokerage brokerage, List<Symbol> symbols, DateTime
+                rows = data;
             }
-        }
 
-        /// <summary>
-        /// Write out the data in LEAN format (daily or hour resolutions)
-        /// </summary>
-        /// <param name="source">IEnumerable source of the data: sorted from oldest to newest.</param>
-        /// <remarks>This function performs a merge (insert/append/overwrite) with the existing Lean zip file</remarks>
-        private void WriteDailyOrHour(IEnumerable<BaseData> source)
-        {
-            var lines = new List<string>();
-            var lastTime = new DateTime();
-
-            // Determine file path
-            var outputFile = GetZipOutputFileName(_dataDirectory, lastTime);
-
-            // Load new data rows into a SortedDictionary for easy merge/update
-            var newRows = new SortedDictionary<DateTime, string>(source.ToDictionary(x => x.Time, x => LeanData.GenerateLine(x, _securityType, _resolution)));
-            SortedDictionary<DateTime, string> rows;
-
-            if (File.Exists(outputFile))
+            if (!_appendToZips && fileExists)
             {
-                // If file exists, we load existing data and perform merge
-                rows = LoadHourlyOrDailyFile(outputFile);
-                foreach (var kvp in newRows)
-                {
-                    rows[kvp.Key] = kvp.Value;
-                }
+                File.Delete(filePath);
+                Log.Trace("LeanDataWriter.Write(): Existing deleted: " + filePath);
             }
-            else
+
+            // If our file doesn't exist its possible the directory doesn't exist, make sure at least the directory exists
+            if (!fileExists)
             {
-                // No existing file, just use the new data
-                rows = newRows;
+                Directory.CreateDirectory(Path.GetDirectoryName(filePath));
             }
 
-            // Loop through the SortedDictionary and write to file contents
-            foreach (var kvp in rows)
+            if (_appendToZips)
             {
-                // Build the line and append it to the file
-                lines.Add(kvp.Value);
-            }
+                var bytes = Encoding.UTF8.GetBytes(string.Join("\n", rows.Values));
+                _dataCacheProvider.Store($"{filePath}#{entryName}", bytes);
 
-            // Write the file contents
-            if (lines.Count > 0)
+                Log.Trace($"LeanDataWriter.Write(): Appended: {filePath} @ {entryName}");
+            }
+            else
             {
-                WriteFile(outputFile, lines, lastTime);
+                // Write out this data string to a zip file
+                var tempFilePath = filePath + ".tmp";
+                Compression.ZipData(tempFilePath, entryName, rows.Values);
+
+                // Move temp file to the final destination with the appropriate name
+                File.Move(tempFilePath, filePath);
+                Log.Trace("LeanDataWriter.Write(): Created: " + filePath);
```

Comment:
Think ideally should also use the `_dataCacheProvider.Store` here? Think we already handled the `_appendToZips` behavior deleting the file if it existed?

Reply:
I think this is unique here, in that it would replace the entire zip. Not just the entry in the zip. Not sure where or why we would use it this way 🤔

Reply:
Solved 👍🏽  _appendToZip is not necessary. 

There is no case in which we will want to delete the zip rather than just replace the entry. By replacing the entry we are deleting the entry itself and writing a new one in its place. Leave the zip alone.

This also address issues with the DataCacheProvider having the file open when this part would attempt to delete. 

End result is to let the DataCacheProvider handle the writing process entirely, do not attempt to modify files in LeanDataWriter.